### PR TITLE
Add VS Code extension article

### PR DESCRIPTION
**Pull request by ivanbuhov**

### CODE REVIEW

File: core-concepts/visual-studio-code-extension.md [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+To install the [NativeScript extension for Visual Studio Code](https://marketplace.visualstudio.com/items?itemName=Telerik.nativescript) open the Command Palette (`F1` or `Cmd+Shift+P`) and run `Extensions: Install Extension` command, then search for 'NativeScript' and choose it from the list.
+
+![Installing NativeScript extension for Visual Studio Code](../img/visual-studio-code-extension/install.png)
+
+If you have installed it successfully, it will appear in the list of installed extensions. You can see it if you run `Extensions: Show Installed Extensions` command from the Command Palette.
+
+## Debugging
+
+Open your application folder in VS Code (your application folder is the one created by `tns create` command, not your `app` folder).
```
Comment (ErjanGavalji) [inline core-concepts/visual-studio-code-extension.md]: This is a bit confusing. If I got it correctly, how about this wording:

Open the `app` folder of your NativeScript project, created via the `tns create` command

File: core-concepts/visual-studio-code-extension.md [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+### Launch an application with debugger
+
+Choose one of the launch configurations (e.g. `Launch on iOS Emulator`) and press the `Start` button next to the menu. This will run your app on the iOS emulator and attach the VS Code debugger. The app should break on the first JavaScript/TypeScript statement. More information about the debugging support in VS Code can be found in the [VS Code Debugging Guide](https://code.visualstudio.com/docs/editor/debugging).
+
+![NativeScript Debugging](../img/visual-studio-code-extension/nativescript-debugging.png)
+
+### Attach the debugger to already running app
+
+If you have already running NativeScript application on your Android/iOS emulator/device you can attach the VS Code debugger to it, without rebuilding and restarting the application. Just choose the right attach configuration from the debug configuration menu, depending on the platform and the target of the app, and press the start button.
```
Comment (ErjanGavalji) [inline core-concepts/visual-studio-code-extension.md]: I would use "If you already have a NativeScript application running on..."

and

"Just select the right attach configuration through the debug configuration menu and press the start button. The attach configuration depends on the app platform and its version.

File: core-concepts/visual-studio-code-extension.md [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+### TypeScript support
+
+In case you are writing your app in TypeScript you have fully functional debugging support in Visual Studio Code. To debug TypeScript, make sure the TS compiler is producing source maps by setting `"sourceMap": true"` in your `tsconfig.json`.
+
+However, instead of manually setuping TS compiler options and TS build tasks we strongly recommend using the [TypeScript plugin for NativeScript](https://github.com/NativeScript/nativescript-dev-typescript) which handles all configuration settings and garantees smooth integration with Visual Studio Code.
```
Comment (ErjanGavalji) [inline core-concepts/visual-studio-code-extension.md]: I would make this an indent block, like:

> Instead of manually setting up the TypeScript compiler options and build tasks, using the [TypeScript plugin for NativeScript](https://github.com/NativeScript/nativescript-dev-typescript) is strongly recommended. The plugin handles the creation of all the configuration settings and guarantees smooth integration to Visual Studio Code.

File: core-concepts/visual-studio-code-extension.md [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+### Supply additional arguments to the debug command
+
+Starting a particular debug configuration, under the hood executes `tns debug` command with various arguments. You can append additional arguments by supplying them in the `tnsArgs` property of a debug configuration definition in `launch.json`. For example if you add `"tnsArgs": "--log=trace"` in the `Launch on iOS Emulator` configuration, under the hood VS Code will execute `tns debug ios --emulator --no-client --log=trace` command, which will give you more verbose information in the Debug Console.
```
Comment (ErjanGavalji) [inline core-concepts/visual-studio-code-extension.md]: Consider rewording this like:
Under the hood, starting a particular debug configuration executes the `tns debug`...

The second mention of "under the hood" can be changed to "in the background" to avoid duplication.

**Comment by ErjanGavalji** [comment]
Great job!

:+1: after addressing the comments and having @ikoevska's review

File: core-concepts/visual-studio-code-extension.md (hunk 1 above)
Comment (ivanbuhov) [inline core-concepts/visual-studio-code-extension.md]: Running `tns create MyApp` will create `MyApp` folder and `app` folder in it. The `MyApp` folder is the one to be opened in VS Code, not the `app` folder. IMO, the phrase "Open the `app` folder of your NativeScript project" is a bit confusing, too. What about:

Open your application root folder, created via the `tns create` command.

File: core-concepts/visual-studio-code-extension.md (hunk 2 above)
Comment (ivanbuhov) [inline core-concepts/visual-studio-code-extension.md]: With "...depending on the platform and the target of the app..." I mean that a configuration exists for every platform (android/ios) and every target (device/emulator), but it is obvious and I prefer removing this sentence However, I will use your variant of the first sentence:

"Just select the right attach configuration through the debug configuration menu and press the start button."

File: core-concepts/visual-studio-code-extension.md (hunk 3 above)
Comment (ivanbuhov) [inline core-concepts/visual-studio-code-extension.md]: :+1:

File: core-concepts/visual-studio-code-extension.md (hunk 4 above)
Comment (ivanbuhov) [inline core-concepts/visual-studio-code-extension.md]: :+1:

**Comment by ivanbuhov** [comment]
I will squash all the commits before merging the branch.

File: core-concepts/debugging.md [hunk 5]
```diff
@@ -8,10 +8,11 @@ previous_url: /debugging
 
 # Debugging
 
-You can debug apps developed with the NativeScript framework from the NativeScript CLI.
+You can debug apps developed with the NativeScript framework from both NativeScript CLI and [Visual Studio Code](https://code.visualstudio.com/).
```
Comment (ikoevska) [inline core-concepts/debugging.md]: both the NativeScript CLI and ...

File: core-concepts/debugging.md [hunk 6]
```diff
@@ -49,3 +50,7 @@ For more information about Android debugging, run the following command:
 ```Bash
 tns help debug ios
 ```
+
+## Debugging with Visual Studio Code
+
+To debug NativeScript applications in [Visula Studio Code](https://code.visualstudio.com/) you need the [NativeScript extension for VSCode](https://marketplace.visualstudio.com/items?itemName=Telerik.nativescript). Detailed instructions how to install and setup the extension you can find [here](./visual-studio-code-extension.md).
```
Comment (ikoevska) [inline core-concepts/debugging.md]: - Visula -> Visual
- comma after `[Visula Studio Code](https://code.visualstudio.com/)`
- VSCode -> VS Code
- setup -> set up

Also, you might want to use slug format for the link to the NS doc set. In your case, this should be done like this: `[here]({% slug nativescript-extension-for-visual-studio-code %)`

File: core-concepts/visual-studio-code-extension.md [hunk 7]
```diff
@@ -0,0 +1,79 @@
+---
+title: NativeScript Extension for Visual Studio Code
+description: How to install, setup and use the NativeScript extension for Visual Studio Code.
```
Comment (ikoevska) [inline core-concepts/visual-studio-code-extension.md]: set up

File: core-concepts/visual-studio-code-extension.md [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+position: 11
+slug: nativescript-extension-for-visual-studio-code
+---
+
+# NativeScript Extension for Visual Studio Code
+
+0. [Installation](#installation)
+2. [Debugging](#debugging)
```
Comment (ikoevska) [inline core-concepts/visual-studio-code-extension.md:11]: Something's wrong with the numbering here. This should work fine in the output but consider fixing the source.

File: core-concepts/visual-studio-code-extension.md [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+  1. [Generate launch configurations](#generate-launch-configurations)
+  2. [Launch an application with debugger](#launch-an-application-with-debugger)
+  3. [Attach the debugger to already running app](#attach-the-debugger-to-already-running-app)
```
Comment (ikoevska) [inline core-concepts/visual-studio-code-extension.md]: to an already

fix here, in the anchor address and in the title below.

File: core-concepts/visual-studio-code-extension.md [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+  4. [TypeScript support](#typescript-support)
```
Comment (ikoevska) [inline core-concepts/visual-studio-code-extension.md]: (Not a mandatory change) Consider rephrasing this point to ensure parallelism in the phrasing of the bullet points (that is that all should be structured and sound similarly). In this case, perhaps, a good substitution would be "Use TypeScript" or "Develop with TypeScript" or something similar, beginning with a verb :)

File: core-concepts/visual-studio-code-extension.md [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+  5. [Supply additional arguments to the debug command](#supply-additional-arguments-to-the-debug-command)
+  6. [Turn on debugging protocol diagnostic logging](#turn-on-diagnostic-logging)
+3. [Use NativeScript specific commands](#use-nativescript-specific-commands)
+4. [NativeScript CLI Version Detection](#nativescript-cli-version-detection)
+
+## Installation
+
+To install the [NativeScript extension for Visual Studio Code](https://marketplace.visualstudio.com/items?itemName=Telerik.nativescript) open the Command Palette (`F1` or `Cmd+Shift+P`) and run `Extensions: Install Extension` command, then search for 'NativeScript' and choose it from the list.
+
+![Installing NativeScript extension for Visual Studio Code](../img/visual-studio-code-extension/install.png)
```
Comment (ikoevska) [inline core-concepts/visual-studio-code-extension.md]: Installing the ...

File: core-concepts/visual-studio-code-extension.md [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+
+If you have installed it successfully, it will appear in the list of installed extensions. You can see it if you run `Extensions: Show Installed Extensions` command from the Command Palette.
```
Comment (ikoevska) [inline core-concepts/visual-studio-code-extension.md]: After the installation completes, the extension appears in the list of installed extensions.

File: core-concepts/visual-studio-code-extension.md [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+
+## Debugging
+
+Open your application root folder, created via the `tns create` command, in Visual Studio Code.
```
Comment (ikoevska) [inline core-concepts/visual-studio-code-extension.md]: created with/created by

File: core-concepts/visual-studio-code-extension.md [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+
+### Generate launch configurations
+
+Click the debugging icon ![VS Code debug panel](../img/visual-studio-code-extension/debug-panel.png) in the View bar, and then click the gear icon ![gear icon](../img/visual-studio-code-extension/gear-icon.png) to choose the NativeScript debug environment. A `launch.json` file should be generated in your `.vscode` folder located next to the `app` folder. The configurations, described in `launch.json` can be selected from the menu in the Debug Panel.
+
+![Installing NativeScript extension for Visual Studio Code](../img/visual-studio-code-extension/configurations-menu.png)
+
+### Launch an application with debugger
+
+Choose one of the launch configurations (e.g. `Launch on iOS Emulator`) and press the `Start` button next to the menu. This will run your app on the iOS emulator and attach the VS Code debugger. The app should break on the first JavaScript/TypeScript statement. More information about the debugging support in VS Code can be found in the [VS Code Debugging Guide](https://code.visualstudio.com/docs/editor/debugging).
```
Comment (ikoevska) [inline core-concepts/visual-studio-code-extension.md]: run your app in the iOS emulator

File: core-concepts/visual-studio-code-extension.md [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+
+![NativeScript Debugging](../img/visual-studio-code-extension/nativescript-debugging.png)
+
+### Attach the debugger to already running app
+
+If you already have a NativeScript application running on your Android/iOS emulator/device you can attach the VS Code debugger to it, without rebuilding and restarting the application. Just select the right attach configuration through the debug configuration menu and press the start button.
```
Comment (ikoevska) [inline core-concepts/visual-studio-code-extension.md]: "Android/iOS emulator/device" is a bit ambiguous when reading it. Consider making it "on your Android or iOS emulator or device"

"the right attach configuration" is not entirely clear. Consider making this "the correct attach configuration" or "the desired attach configuration"

File: core-concepts/visual-studio-code-extension.md [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+
+### TypeScript support
+
+In case you are writing your app in TypeScript you have fully functional debugging support in Visual Studio Code. To debug TypeScript, make sure the TS compiler is producing source maps by setting `"sourceMap": true"` in your `tsconfig.json`.
+
+> Instead of manually setting up the TypeScript compiler options and build tasks, using the [TypeScript plugin for NativeScript](https://github.com/NativeScript/nativescript-dev-typescript) is strongly recommended. The plugin handles the creation of all the configuration settings and guarantees smooth integration to Visual Studio Code.
```
Comment (ikoevska) [inline core-concepts/visual-studio-code-extension.md]: Switch the places of the two sub-sentences :)

"Using the .... recommended instead of manually setting up the..."

Comment (ikoevska) [inline core-concepts/visual-studio-code-extension.md]: smooth integration to -> smooth integration with

File: core-concepts/visual-studio-code-extension.md [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+
+Under the hood, starting a particular debug configuration executes the `tns debug` command with various arguments. You can append additional arguments by supplying them in the `tnsArgs` property of a debug configuration definition in `launch.json`. For example if you add `"tnsArgs": "--log=trace"` in the `Launch on iOS Emulator` configuration, in the background VS Code will execute `tns debug ios --emulator --no-client --log=trace` command, which will give you more verbose information in the Debug Console.
+
+![Supply custom arguments to the debug command](../img/visual-studio-code-extension/nativescript-tns-args.png)
+
+### Turn on diagnostic logging
+
+If the `diagnosticLogging` flag for a particular debug configuration in `launch.json` file is set to `true` (its default value is `false`), diagnostic messages will be logged on the debugger console after attaching the debugger to the NativeScript application. This is useful for rough debugging of the extension itself, because all sent/received messages which are part of the communication between the frontend and the backend are logged.
+
+## Use NativeScript specific commands
```
Comment (ikoevska) [inline core-concepts/visual-studio-code-extension.md]: NativeScript-specific (for all instances)

File: core-concepts/visual-studio-code-extension.md [hunk 18]
```diff
@@ -0,0 +1,79 @@
...
+
+Type `nativescript` in the Command Palette and you will see all NativeScript specific commands. Currently there are only two of them but the list will grow in the future.
+
+![NativeScript commands](../img/visual-studio-code-extension/nativescript-commands.png)
+
+Run command is the equivalent to `tns run` in the NativeScript CLI. It allows you to build, deploy and run your app on emulator/device directly from Visual Studio Code.
```
Comment (ikoevska) [inline core-concepts/visual-studio-code-extension.md]: It lets you build...

File: core-concepts/visual-studio-code-extension.md [hunk 19]
```diff
@@ -0,0 +1,79 @@
...
+
+## NativeScript CLI Version Detection
+
+The extension is depending on a globally installed NativeScript CLI. It will show an error message if it can't find it.
+
+![NativeScript not found](../img/visual-studio-code-extension/nativescript-not-found-error-message.png)
+
+> The extension is supposed to work with a specific NativeScript CLI version and if you have another version installed you will see a warning message. You can still use it with the inconsistent version of the CLI and it's likely to work seamlessly but we recommend updating either the NativeScript CLI or the VS Code extension.
```
Comment (ikoevska) [inline core-concepts/visual-studio-code-extension.md]: - The extension is supposed to work with a specific -> The extension requires a specific
- comma after "if you have another version installed"
- You can still use it with the inconsistent version of the CLI and it's likely to work seamlessly but we recommend updating either the NativeScript CLI or the VS Code extension. -> The extension is likely to work with the unsupported version but it is recommended that you update the NativeScript CLI or the VS Code extension.

**Comment by ikoevska** [comment]
Great work, :+1: after addressing my comments.
